omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,899 @@
+                return std::make_pair(nullptr, 0);
+            }
+        }
+    }
+};
+
+
+/**
+ * Performs an unordered breadth-first search,
+ * thereby building a rooted level structure.
+ */
+template <typename IndexType>
+void ubfs(std::shared_ptr<const OmpExecutor> exec, const size_type num_vertices,
+          const IndexType *const row_ptrs, const IndexType *const col_idxs,
+          IndexType *const
+              levels,  // Must be inf/max in all nodes connected to source
+          const IndexType start, const IndexType max_degree)
+{
+    const auto max_threads = omp_get_max_threads();
+
+    // This is an upper bound for how many nodes may be in the work set
+    // at maximum at any time.
+    const auto work_bound = (size_type)max_threads * num_vertices;
+    UbfsLinearQueue<IndexType> q(exec, work_bound);
+
+    // Start from the given start node.
+    q.enqueue_chunk(&start, 1);
+    levels[start] = 0;
+
+    // Used for synchronization of (pseudo)remaining work.
+    int threads_working = 0;
+
+#pragma omp parallel firstprivate(max_degree)
+    {
+        // This is an upper bound, as '.deque_chunk...' is upper bounded by
+        // chunk_bound.
+        gko::vector<IndexType> local_to_insert(max_degree * chunk_bound, exec);
+        auto local_to_insert_size = 0;
+
+        while (true) {
+            // Get some nodes from the locked q.
+            const auto pair = q.dequeue_chunk(&threads_working);
+            const auto chunk = pair.first;
+            const auto chunk_size = pair.second;
+
+            // If there are nodes left to process.
+            // If there are no nodes returned that also means there are no nodes
+            // still in flight.
+            if (chunk_size > 0) {
+                // For each node in the chunk.
+                for (auto chunk_i = 0; chunk_i < chunk_size; ++chunk_i) {
+                    const auto node = chunk[chunk_i];
+
+                    IndexType level;
+#pragma omp atomic read
+                    level = levels[node];
+
+                    const auto new_neighbour_level = level + 1;
+
+                    // Go through neighbours.
+                    const auto row_start = row_ptrs[node];
+                    const auto row_end = row_ptrs[node + 1];
+
+                    for (auto neighbour_i = row_start; neighbour_i < row_end;
+                         ++neighbour_i) {
+                        const auto neighbour = col_idxs[neighbour_i];
+
+                        IndexType old_neighbour_level;
+                        do {
+#pragma omp atomic read
+                            old_neighbour_level = levels[neighbour];
+
+                            const auto is_new_smaller =
+                                new_neighbour_level < old_neighbour_level;
+
+                            if (!is_new_smaller) {
+                                goto continue_outer;
+                            }
+                        } while (!__atomic_compare_exchange_n(
```

Comment:
That would require `levels` being all `std::atomic`s, sadly you can't just wrap a value / reference to it with a `std::atomic` and use that for one operation. 
I did not want to do that for two reasons:
1. OpenMP doesn't specify how it interacts with `std::atomic`, a judicious use of it might result in undefined behavior for some compilers (yes, OpenMP also doesn't specify anything about compiler intrinsics, I consider this one call to be less "attack surface" though).
2. That would mean all memory operations there must be atomic, however that is simply not necessary.

Also, this, contrary to its appearance, seems to be portable. GCC even seems to compile it just fine for ARM.

And all of this just because OpenMP doesn't feel like providing a CAS operation.

Reply:
Fair points, seems like [N4013](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4013.html) is still pending?

Reply:
I was not able to find any information on the current status of that proposal. Might have been discarded. But yes, I double checked - no way to do that in defined, sound C++ currently.